Name inference for min(Tensor, dim?) / max(Tensor, dim?)

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26563 Named tensor support for logsumexp, mode, kthvalue, median
* #26558 Wrap dimensions during named inference
* **#25582 Name inference for min(Tensor, dim?) / max(Tensor, dim?)**

There are a lot of min/max overloads. This PR adds name inference to
the following overloads for (both) min and max:
- min(Tensor, int dim)
- min(Tensor, Dimname dim)
- min(Tensor)  (full reduction)

Test Plan:
- new tests [namedtensor ci]

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25582

Differential Revision: [D17521607](https://our.internmc.facebook.com/intern/diff/D17521607)

zou3519 commented:
@pytorchbot retest this please

nairbv commented on aten/src/ATen/native/TensorCompare.cpp:
```diff
@@ -178,6 +181,16 @@ std::tuple<Tensor &,Tensor &> max_out(Tensor& max, Tensor& max_indices,
+  auto result = max_out_unnamed(max, max_indices, self, dim, keepdim);
+#ifdef BUILD_NAMEDTENSOR
+  namedinference::propagate_names_for_reduction(max, self, dim, keepdim);
+  namedinference::propagate_names_for_reduction(max_indices, self, dim, keepdim);
```
It doesn't seem right for the dim names to be the same for values and indices.

If I have a tensor with dimensions X, Y, and I take the max along dimension Y, the resulting values will still have dimension X since Y has been reduced, but the indices will be along dimension Y.  I.e. If I wanted to use the index to look up the max value in the original tensor, the dimension I'd be looking for would be the Y dimension.

zou3519 replied:
I disagree. Here's a concrete example. Let `img = torch.randn(2, 3, 5, names=('C', 'H', 'W')` be an image.

```
values, indices = img.max('C')
```

`indices.size()` is `3, 5` and corresponds to the H and W dimensions. Given concrete `h` and `w, the value inside `c = indices[h, w]` tells us the index `c` such that `img[c, h, w]` is maximized.

nairbv replied:
ah, that's also a reasonable way of thinking of it. I think both potentially work? I was thinking what dimension the index references in the original tensor but maybe that makes less sense with more dimensions.

It might be helpful to have an explicit test that shows how this behavior works and/or some extra explanation in the docs, since it may be non-obvious to other users.

zou3519 replied:
> I was thinking what dimension the index references in the original tensor but maybe that makes less sense with more dimensions.

I think there's two things that are being confused here. (1) dimensions are named, individual values don't have an associated "name" or "named dimension". (2) the indices tensor contains values, each of which represent an index to the dimension that got reduced over (C in the above example). 

> It might be helpful to have an explicit test that shows how this behavior works and/or some extra explanation in the docs, since it may be non-obvious to other users.

I'm not actually sure how users use `indices` after getting it out of `max`. It's kinda difficult to use. I'll go hunting for some code samples out in the wild.

nairbv replied:
I get that it's the dimensions that are named, but an index is indexing into a particular dimension.

One can reconstruct the values tensor the index and original tensor like this:
```
t=torch.tensor([[5, 1, 1,],[1,1,5]])
a,b = torch.max(t, 0)
a,b
(tensor([5, 1, 5]), tensor([0, 1, 1]))
[t[b[v], v] for v in range(0, t.size(1))]
```
In this example `b[v]` provides the index into the 0'th dimension of `t`, which is why I was thinking that b's dimension name would be the same as the name of the (removed) 0th dimension of `t`.

zou3519 replied:
In your example, `b[v]` is a scalar tensor. Scalar tensors have no names. It doesn't matter what the names of `b` are, because when the indexing is performed, you are ultimately indexing with scalar tensors.

Now, this logic can be different if there's a concrete example of indexing with a tensor with names, but I need to go think about that.

zou3519 replied:
Here's a generalized example for how to reconstruct `values` from `indices` and `tensor`:
```
import torch

tensor = torch.randn(2, 3, 5)
values, indices = tensor.max(1)

# to reconstruct values from tensor & indices
tensor.gather(1, indices.unsqueeze(1)).squeeze(1)
```
I don't have a named inference rule implemented for `gather`. However, it is probably going to be: "indices.names must match tensor.names" because that's intuitively what gather requires. Given that, indices should have the reduced names for this example to work out.

nairbv replied:
another way of looking at it:
```
In [94]: t = torch.rand(3,3,3)
In [95]: a, b = t.max(0)
In [96]: t[b[0, 0], 0, 0] == a[0, 0]
Out[96]: tensor(True)
In [97]: t[b[1, 2], 1, 2] == a[1, 2]
Out[97]: tensor(True)
```
As I write that out, it is a bit clearer to me that the indices/dimensions I'm referencing in `b` match `a`.

I think it'll be helpful to document the behavior though and/or have an explicit test showing how we expect it to behave.

zou3519 replied:
> As I write that out, it is a bit clearer to me that the indices/dimensions I'm referencing in b match a.

Sorry, I'm having trouble following your example and what it demonstrates. Could you explain it in a little more detail?

> I think it'll be helpful to document the behavior though

Documentation for named tensors coming soon. Won't be in this PR.

> and/or have an explicit test showing how we expect it to behave.

We do assert that the names on `indices` matches the names on `values` in the test. The reduction test assert that all outputs have names that are equal to the names in the original tensor without the reduced dimension.

zou3519 committed:
Update on "Name inference for min(Tensor, dim?) / max(Tensor, dim?)"


There are a lot of min/max overloads. This PR adds name inference to
the following overloads for (both) min and max:
- min(Tensor, int dim)
- min(Tensor, Dimname dim)
- min(Tensor)  (full reduction)

Test Plan:
- new tests [namedtensor ci]

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25582

Differential Revision: [D17185069](https://our.internmc.facebook.com/intern/diff/D17185069)

zou3519 committed:
Update on "Name inference for min(Tensor, dim?) / max(Tensor, dim?)"


There are a lot of min/max overloads. This PR adds name inference to
the following overloads for (both) min and max:
- min(Tensor, int dim)
- min(Tensor, Dimname dim)
- min(Tensor)  (full reduction)

Test Plan:
- new tests [namedtensor ci]

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25582

Differential Revision: [D17185069](https://our.internmc.facebook.com/intern/diff/D17185069)

zou3519 commented:
@nairbv is this good to go? I'm planning on addressing your comment on the documentation when I write documentation for named tensors in a batch.

zou3519 committed:
Update on "Name inference for min(Tensor, dim?) / max(Tensor, dim?)"


There are a lot of min/max overloads. This PR adds name inference to
the following overloads for (both) min and max:
- min(Tensor, int dim)
- min(Tensor, Dimname dim)
- min(Tensor)  (full reduction)

Test Plan:
- new tests [namedtensor ci]

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25582

Differential Revision: [D17185069](https://our.internmc.facebook.com/intern/diff/D17185069)

zou3519 committed:
Update on "Name inference for min(Tensor, dim?) / max(Tensor, dim?)"


There are a lot of min/max overloads. This PR adds name inference to
the following overloads for (both) min and max:
- min(Tensor, int dim)
- min(Tensor, Dimname dim)
- min(Tensor)  (full reduction)

Test Plan:
- new tests [namedtensor ci]

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25582

Differential Revision: [D17185069](https://our.internmc.facebook.com/intern/diff/D17185069)
